packages/artifact/src/internal/shared/util.ts

```diff
@@ -69,3 +70,170 @@ export function getBackendIdsFromToken(): BackendIds {
 
   throw InvalidJwtError
 }
+
+/**
+ * Masks the `sig` parameter in a URL and sets it as a secret.
+ * @param url The URL containing the `sig` parameter.
+ * @returns A masked URL where the sig parameter value is replaced with '***' if found,
+ *          or the original URL if no sig parameter is present.
+ */
+export function maskSigUrl(url: string): string {
+  if (!url) return url
+
+  try {
+    const rawSigRegex = /[?&](sig)=([^&=#]+)/gi
+    let match
+
+    while ((match = rawSigRegex.exec(url)) !== null) {
+      const rawSignature = match[2]
+      if (rawSignature) {
+        setSecret(rawSignature)
+      }
+    }
+
+    let parsedUrl: URL
+    try {
+      parsedUrl = new URL(url)
+    } catch {
+      try {
+        parsedUrl = new URL(url, 'https://example.com')
+      } catch (error) {
+        debug(`Failed to parse URL: ${url}`)
+        return maskSigWithRegex(url)
+      }
+    }
+
+    let masked = false
+    const paramNames = Array.from(parsedUrl.searchParams.keys())
+
+    for (const paramName of paramNames) {
+      if (paramName.toLowerCase() === 'sig') {
+        const signature = parsedUrl.searchParams.get(paramName)
+        if (signature) {
+          setSecret(signature)
+          setSecret(encodeURIComponent(signature))
+          parsedUrl.searchParams.set(paramName, '***')
+          masked = true
+        }
+      }
+    }
+    if (masked) {
+      return parsedUrl.toString()
+    }
+
+    if (/([:?&]|^)(sig)=([^&=#]+)/i.test(url)) {
+      return maskSigWithRegex(url)
+    }
+  } catch (error) {
+    debug(
+      `Error masking URL: ${
+        error instanceof Error ? error.message : String(error)
+      }`
+    )
+    return maskSigWithRegex(url)
+  }
+
+  return url
+}
+
+/**
+ * Fallback method to mask signatures using regex when URL parsing fails
+ */
+function maskSigWithRegex(url: string): string {
+  try {
+    const regex = /([:?&]|^)(sig)=([^&=#]+)/gi
+
+    return url.replace(regex, (fullMatch, prefix, paramName, value) => {
+      if (value) {
+        setSecret(value)
+        try {
+          setSecret(decodeURIComponent(value))
+        } catch {
+          // Ignore decoding errors
+        }
+        return `${prefix}${paramName}=***`
+      }
+      return fullMatch
+    })
+  } catch (error) {
+    debug(
+      `Error in maskSigWithRegex: ${
+        error instanceof Error ? error.message : String(error)
+      }`
+    )
+    return url
+  }
+}
+
+/**
+ * Masks any URLs containing signature parameters in the provided object
+ * Recursively searches through nested objects and arrays
+ */
+export function maskSecretUrls(
```

Comment:
If we're not using this right now, it might be better to not introduce this

Reply:
yup made it less complex